teps/0073-simplify-metrics.md

```diff
@@ -0,0 +1,427 @@
+
+### Setting Level of Metrics for TaskRun or PipelineRun
+
+We can add a `config-observability` option to switch between `TaskRun` and
+`PipelineRun` level metrics, `Task` and `Pipeline` or namespace level. 
+`metrics.taskrun.level` and `metrics.pipelinerun.level` fields will indicate
+at what level to aggregate metrics. 
+- When they are set to `namespace`, they will remove `Task`, `TaskRun` and,
+  `PipelineRun` and `TaskRun` label respectively in metrics.
+- When set to `task` or `pipeline`, they will remove `TaskRun` and `PipelineRun` 
+  label respectively. 
+- When set to `taskrun` or `pipelinerun`, current behaviour will be exhibited.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: config-observability
+  namespace: tekton-pipelines
+  labels:
+    app.kubernetes.io/instance: default
+    app.kubernetes.io/part-of: tekton-pipelines
+data:
+  metrics.task.level: "task"
+```
+
+```log
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="anonymous",taskrun="duplicate-pod-task-run-ytqrdxja",le="10"} 1
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="anonymous",taskrun="duplicate-pod-task-run-ymelobwl",le="10"} 1
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="anonymous",taskrun="duplicate-pod-task-run-xnuasulj",le="10"} 1
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="test",taskrun="duplicate-pod-task-run-tqerstbj",le="10"} 1
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="test",taskrun="duplicate-pod-task-run-alcdjfnk",le="10"} 1
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="test",taskrun="duplicate-pod-task-run-rtyjsdfm",le="10"} 1
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",status="success",task="test",taskrun="duplicate-pod-task-run-iytyhksd",le="10"} 1
+```
+
+When the option is `task`, these metrics will be merged into two based on `task` label.
+
+```log
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",task="anonymous",status="success",le="10"} 3
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw",task="test", status="success",le="10"} 4
+```
+
+When the option is `namespace`, these metrics will be merged into one.
+
+```log
+tekton_taskrun_duration_seconds_bucket{namespace="arendelle-nsfqw","status="success",le="10"} 7
+```
+
+### Change metrics type
+
+Large metrics volume is mostly caused by histogram metrics. One single histogram metrics will produce 15 metrics. These metrics type could be changed to gauge which would reduce 15 metrics to one.
+For example, metrics like `tekton_pipelinerun_duration_seconds`, `taskrun_duration_seconds`, `tekton_pipelinerun_taskrun_duration_seconds` are histogram. But the metrics could not provide much information at `TaskRun` level and can be changed to gauge.
```

Comment:
As per my previous comment, I would advise removing them. The problem with these metrics can be seen in the `Before` example you've shared below, these metrics only record the duration of one series, so it is neither meaningful to have percentiles based on it nor averages.

Reply:
Yes, removing `pipelinerun` and `taskrun` labels altogether make sense. We get an additional metrics because of these labels with only one update. This data can be computed using logging stack.

Reply:
I agree! Data on how long it took to execute a specific `PipelineRun` should be taken from `tektoncd/results` if needed.